cogs/strawpoll.py

```diff
@@ -0,0 +1,292 @@
+import asyncio
+import os
+from collections import defaultdict
+from .utils import checks
+from .utils.dataIO import dataIO
+from concurrent.futures import CancelledError
+from html import unescape
+from time import sleep, time, strftime
+from discord.ext import commands
+
+def _get_poll(poll_id):
+    # https://strawpoll.me/api/v2/polls/{poll_id}
+    https = urllib3.PoolManager(
+        cert_reqs='CERT_REQUIRED',
+        ca_certs=certifi.where())
+    try:
+        request = https.request( 'GET', 
+            'https://strawpoll.me/api/v2/polls/' + str(poll_id),
+            retries=3, timeout=3.0)
+        json_response = json.loads(request.data.decode('utf-8'))
+    except urllib3.exceptions.HTTPError:
+        return None
+    except json.decoder.JSONDecodeError:
+        return None
+    return json_response
+    
+def _post_poll(title, options, multi):
+    # build json request
+    data = {
+        'title': title,
+        'options': options,
+        'multi': multi,
+        'dupcheck': 'normal',
+        'captcha': False }
+    encoded_data = json.dumps(data).encode('utf-8')
+    # urllib3 pool manager in https mode
+    https = urllib3.PoolManager(
+        cert_reqs='CERT_REQUIRED',
+        ca_certs=certifi.where())
+    try:
+        request = https.request(
+            'POST', 'https://strawpoll.me/api/v2/polls',
+            body=encoded_data, headers={
+                'Content-Type': 'application/json',
+                'Accept-Charset': 'utf-8'
+            },
+            retries=3, timeout=3.0)
+        json_response = json.loads(request.data.decode('utf-8'))
+    except urllib3.exceptions.HTTPError:
+        return None
+    return json_response
+
+class _Poll:
+    """Internal poll class for Strawpoll cog"""
+    
+    def __init__(self, bot, settings, message, author, poll_id, poll_length):
+        self.bot = bot
+        self.settings = settings
+        self.message = message
+        self.author = author
+        self.poll_id = poll_id
+        self.poll_length = poll_length
+        self.start_time = time()
+
+    def _bar_creator(self, data, option):
+        display_bar = " :: "
+        bar_length = self.settings['bar_length']
+        for y in range(bar_length):
+            if (max(data['votes']) > 0):
+                pct = float(
+                    data['votes'][option])/float(max(data['votes']))
+            else:
+                continue
+            if (pct != 0 and pct >= float(y)/float(bar_length-1)):
+                display_bar += '|'
+        return display_bar
+
+    async def update_results(self):
+        data = _get_poll(self.poll_id)
+        if data is None:
+            await self.bot.edit_message(self.message, 
+                embed=discord.Embed(title="Error receiving strawpoll data!"))
+        embed=discord.Embed(title=unescape(data['title']), 
+            url='https://strawpoll.me/' + str(self.poll_id), 
+            description="Strawpoll Results", color=self.author.color)
```

Comment:
I really like the use of color here. Not something I think anyone will notice but it's such a good idea.

Reply:
Thanks! I think I saw mee6 or some other bot do it a while back though, so I can't claim it as an original idea.